### DISCUSSION TITLE
Rule to send MQTT when A0 less than a value, as fast as possible!

### DISCUSSION THREAD

**Discussion by SimonFili** [arendst/Tasmota#21394] [Q&A]
Hi all, been trying for 2 hours many combinations of rules, backlog to do this, but nothing works.

So my goal is to have a water level sensor.  Pretty easy, +3.3V to a 1 Mohm resistor to wire1 and wire2 to ground.  A0 is connected between the resistor and wire1.  So when I put the 2 wires in water, A0 goes below 2000 in the analog reading.

I want to use this input to stop a pump ASAP, controlled through Home Assistant with MQTT messaging.

Teleperiod being a minimum of 10 seconds, is too slow for my use case.

So I'm trying to set up a rule that would trigger on 
`Rule1 ON Analog#a0<2000`

Then the DO:
`DO Publish stat/ESP32S3_T1/ALERT "Level1_On" ENDON`

In my GREAT IGNORANCE!  I'm assuming this would trigger as soon as A0 goes below 2000, but it does not work.

Any help would be appreciated.

Yes, I read example here and tried many variations with backlog also. https://tasmota.github.io/docs/Rules

Many thanks for any help!

**Comment by arendst** [comment]
Try ``ON Analog#A0div10<200``

**Reply by sfromis** [reply]
Rules for sensor values are not "as soon as", but based on periodic checking, which for analog also includes time for oversampling. Basically, you don't get "fast" reactions.

**Comment by SimonFili** [comment]
Thanks for the inputs!

Ok, so I have this rule:

`Rule1 ON Analog#A1div10<200 DO Publish stat/ESP32S3_T1/ALERT "Level1High" ENDON
`
My values are from 0 to 4095, when the wires are in water, I see around 800 on the logs values for A1 (I have A2 also)

```
12:48:15.942 MQT: tele/ESP32S3_T1/SENSOR = {"Time":"2024-05-09T12:48:15","ANALOG":{"A1":4085,"A2":4038}}
12:48:45.972 MQT: tele/ESP32S3_T1/SENSOR = {"Time":"2024-05-09T12:48:45","ANALOG":{"A1":832,"A2":4014}}
```

So 800/10 = 80 (smaller than 200), then the rule should be triggered and output the publishing, right?
I assume here that div10 does that ?

It's not working for me! 

Here's the log output for rule1
```
12:51:12.585 CMD: rule1
12:51:12.586 SRC: WebConsole from 192.168.10.100
12:51:12.587 CMD: Grp 0, Cmd 'RULE', Idx 1, Len 0, Pld -99, Data ''
12:51:12.590 MQT: stat/ESP32S3_T1/RESULT = {"Rule1":{"State":"ON","Once":"ON","StopOnError":"OFF","Length":73,"Free":438,"Rules":"ON Analog#A1div10<200 DO Publish stat/ESP32S3_T1/ALERT \"Level1High\" ENDON"}}
```

The only time I see the output is when I enter the line for rule1.
```
12:52:55.984 CMD: Rule1 ON Analog#A1div10<200 DO Publish stat/ESP32S3_T1/ALERT "Level1High" ENDON
12:52:55.985 SRC: WebConsole from 192.168.10.100
12:52:55.986 CMD: Grp 0, Cmd 'RULE', Idx 1, Len 73, Pld -99, Data 'ON Analog#A1div10<200 DO Publish stat/ESP32S3_T1/ALERT "Level1High" ENDON'
12:52:55.988 RUL: Stored uncompressed, would compress from 73 to 69 (-6%)
12:52:55.991 MQT: stat/ESP32S3_T1/RESULT = {"Rule1":{"State":"ON","Once":"ON","StopOnError":"OFF","Length":73,"Free":438,"Rules":"ON Analog#A1div10<200 DO Publish stat/ESP32S3_T1/ALERT \"Level1High\" ENDON"}}
12:52:56.060 RUL: ANALOG#A1DIV10<200 performs "Publish stat/ESP32S3_T1/ALERT "Level1High""
12:52:56.061 SRC: Rule
12:52:56.063 CMD: Grp 0, Cmd 'PUBLISH', Idx 1, Len 34, Pld -99, Data 'stat/ESP32S3_T1/ALERT "Level1High"'
12:52:56.065 MQT: stat/ESP32S3_T1/ALERT = "Level1High"
12:53:05.869 MQT: tele/ESP32S3_T1/STATE = {"Time":"2024-05-09T12:53:05","Uptime":"0T05:25:13","UptimeSec":19513,"Heap":240,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":43,"MqttCount":1,"Berry":{"HeapUsed":4,"Objects":52},"POWER1":"OFF","POWER2":"ON","Wifi":{"AP":1,"SSId":"IoT2.4","BSSId":"A2:DA:C4:A4:46:F2","Channel":1,"Mode":"11n","RSSI":76,"Signal":-62,"LinkCount":1,"Downtime":"0T00:00:03"}}
12:53:05.942 MQT: tele/ESP32S3_T1/SENSOR = {"Time":"2024-05-09T12:53:05","ANALOG":{"A1":4086,"A2":4095}}
```

Even if the sensor is not in water (4086)

**Reply by barbudor** [reply]
You have enabled the Once mode (Rule1 5) so the rule is triggered only once and won't be triggered again until the value goes over 200 to be re-armed and under 200 again to be triggered

**Reply by barbudor** [reply]
If you remove the Once mode (Rule1 4) the the rule will be triggered continuously, overflowing the console and mqtt

**Reply by sfromis** [reply]
A trigger with `DIV10` should not be firing continuously. As the docs says:
> Analog#A0div10   when the A0 input changes by more than 1% it provides a value between 0 and 100

Thus, even less reason to use one-shot

**Reply by SimonFili** [reply]
Ok, I can confirm that rule1 4, triggers continuously, even if the analog value is over 4000.  So I get an artificial trigger?  Not sure if I get the logic.

I saw the doc line 
Analog#A0div10 when the A0 input changes by more than 1% it provides a value between 0 and 100

But I don't understand it.  If I get a change of more than 1% on the input, the output is between 0 and 100?  
So I get A1 at 4090, 1% of this is 40.90.  I still don't get it!  Maybe because I'm French speaking!? ;-)

**Reply by arendst** [reply]
The result of ``Analog#AxDIV10`` is always between minimum 0 and maximum 100.

On the ESP8266 the analog range goes from 0 to 1024. Hence the name ``DIV10``. On ESP32 the analog range changed from 0 to 4096. The name is misleading now but the result is the same; 0 becomes 0, 4096 becomes 100.

So your rule needs to test for a value between 0 and 100. Testing on a value below 200 will always result in a trigger as ALL values of ``DIV10`` are below 200.

**Comment by SimonFili** [comment]
Alright!  Thanks to all of you, it works!  This is the final rule.

`Rule1 ON Analog#A1div10<50 DO Publish stat/ESP32S3_T1/ALERT "Level1High" ENDON; rule1 1`

I added ; rule1 1 to re-enable the rule instead of "rule1 4"

This way, I only get on MQTT on detection.  
With rule1 4, I get a continuous flow of MQTT when water is detected.

So far, with a few tests, the response time seems good for me, inside 1 second!

VERY HAPPY!

Thanks again

**Reply by barbudor** [reply]
I don't know how often this will happen but it's generally a bad idea to turn a rule on and off as everytime the flash settings are save which will tear the flash faster

The Once mode should do the trick or something based on timer

**Reply by SimonFili** [reply]
I see, thanks for this, I will try again with Rule1 4 and see how I can manage multiple level detection inside home assistant.

**Reply by sfromis** [reply]
Unless analog input fluctuations are above 1%, using `div10` should not result in constant messages, like not when the reading is stable.

**Comment by SimonFili** [comment]
Ok so what works for me is this:

```
Rule1 ON Analog#A1div10<50 DO Backlog Var1 %value%; Publish stat/ESP32S3_T1/ALERT "Level1High, Value: %Var1%" ENDON
Rule1 5   # Once : ON
```

This gets me one message when wires touches the water, inside one second.  If I get the wires out of the water, it seems to "reset" it and when I put the wires again in water, I get another message.

So this works perfectly! 

Div10  seems to be more than a simple "mapping" of inputs to get in the 0...100 range but also a form of trigger.

So if I understand this correctly, when using this format "ANALOG#A0div10", there will be a trigger if the input (0 to 4095 for ESP32 and 0 to 1023 on ESP8266) changes for more than 1%.  Then, the output value will be mapped to a value from 0 to 100.

Right ?

**Reply by sfromis** [reply]
You already yourself made the quote about `div10` from the docs:
> when the A0 input changes by more than 1%

It means exactly what it says.

Yes, it is indeed not just a conversion, key words are "when" and "changes". The mapping to 0..100 happens *before* the logic about detecting change. That's also the different thing with this specific value-based trigger; in general, triggers don't care about change or not, unless you save the old value and compare.
